### ISSUE TITLE
VB editor doesn't highlight syntax errors in lambdas

### ISSUE THREAD

**Issue by VBAndCs** [dotnet/roslyn#42857]
In this code:
```VB.NET
Module Program
    Sub Main(args As String())
        Foo(Function() 3,
                Sub(x)
                    Dim x = Function(x) x + 3
                End Sub
        )
    End Sub

    Sub Foo(a As Integer, b As Integer)
             ' Do something
    End Sub

    Sub Foo(a As Func(Of Integer), b As Action(Of Integer))
             ' Do something
    End Sub
End Module
```

The editor puts a red line under `Foo` in the third line ` Foo(Function() 3,` saying there is an overload resolution error (which is not true), and doesn't highlight the real errors involving declaring a local variable x and a param x inside the `Sub(x)` lambda sub that has a param x. These errors are reported in one long message, which may be the reason that the editor can't realize them.
> Overload resolution failed because no accessible 'Foo' can be called with these arguments:    'Public Sub Foo(a As Integer, b As Integer)': Lambda expression cannot be converted to 'Integer' because 'Integer' is not a delegate type.    'Public Sub Foo(a As Integer, b As Integer)': Lambda expression cannot be converted to 'Integer' because 'Integer' is not a delegate type.    'Public Sub Foo(a As Func(Of Integer), b As Action(Of Integer))': Lambda parameter 'x' hides a variable in an enclosing block, a previously defined range variable, or an implicitly declared variable in a query expression.    'Public Sub Foo(a As Func(Of Integer), b As Action(Of Integer))': Variable 'x' is already declared as a parameter of this or an enclosing lambda expression.    'Public Sub Foo(a As Func(Of Integer), b As Action(Of Integer))': Variable declaration without an 'As' clause; type of Object assumed.    'Public Sub Foo(a As Func(Of Integer), b As Action(Of Integer))': Operands of type Object used for operator '+'; runtime errors could occur.

**Comment by CyrusNajmabadi** [comment]
So i get this:

![image](https://user-images.githubusercontent.com/4564579/77812618-b939d300-705f-11ea-8790-e10ab3194e8c.png)

I think what's happening is that there are different potential errors depending on which overload it tries.  So it's letting you know hte potential issues per overload so you can do something about.  

For example, if i get rid of either overload, then i get the specifics shown. i.e.:

![image](https://user-images.githubusercontent.com/4564579/77812647-f605ca00-705f-11ea-8656-6daf9d3d2f02.png)

Or, with the other overload:

![image](https://user-images.githubusercontent.com/4564579/77812657-01f18c00-7060-11ea-8752-f570878e915f.png)

Since there are multiple possibilities here, VB is just letting you know overall that things went wrong and waht the particular problems were with each attempt.

--

I think this is likely by design.

**Comment by VBAndCs** [comment]
If it is by design, then I suggest you should better change this design. I wasted hours trying to get what is wrong as I couldn't read the whole message, because in my actual app the func params were far complex and long. So:
1- divide the long composite message into small distinct messages, to make them easier to read.
2. Highlight the errors in code. Composite messages allow only one pos to highlight.
3. Suppress the overload resolution message.
Thanks

**Comment by CyrusNajmabadi** [comment]
@VBAndCs PRs welcome.

**Comment by jmarolf** [comment]
Error recovery in lambdas has typically been a best-effort heuristic. When lambdas go wrong they can go _really_ wrong and its not always clear what the _best_ error is to show to users.

That being said I think there is always room for improvement and if you can come up with an error-recovery scheme that will make things better for the majority of users we would happily accept it.

**Comment by VBAndCs** [comment]
@jmarolf 
I don't know how Roslyn deal with errors to suggest a solution.
I discovered what is wrong by moving lambdas out of the function call, and assigned them to a delegate variable, hence editor highlighted the errors.
I think it is better to highlight inner errors (in lambdas) first. In nested errors, terminal errors could be the reason of higher level errors. I wouldn't mind to get one true error at first, after solving may appear others or not. In the case here, I tried to fix the overload resolution which drove me away from noticing that the error is in lambda. 
In fact my real code was translated from C#, which for some reason accepts that nested lambdas can have params with the same name, but VB respects the scope roles and give an error, that is hidden by the false overload resolution error.
Seems that the propagated lambda params error is considered an overload resolution failure. Maybe you need to check that the param error isn't a bubbling one.

**Comment by CyrusNajmabadi** [comment]
> I think it is better to highlight inner errors (in lambdas) first.

the problem here is that there are distinct, disjoint sets of errors.  

Note that inner errors *are* highlighted when this is not the case.  You can see that in the images i posted.  This doesn't work in the case you brought up because you'll get a set of disparate, potentially conflicting, issues.

> which for some reason accepts that nested lambdas can have params with the same name,

This is legal in C#.  It is not legal in VB.

**Comment by jmarolf** [comment]
As I said, when lambdas go wrong they tend to go _really_ wrong. @VBAndCs what would your preferred precedence for ordering diagnostics inside lambdas? The problem we have is that most errors inside lambdas cascade outward (as you saw). We could report the inner most error but that might not be the most helpful. It could be something like two conflicting names that is the real issue and that may be reported outside the lambda. Not asking you to solve this problem outright as its very complex and I personally can't think of any obvious wins here. But maybe if we break down the specific situation we can come up with a heuristic that could make things better in a few narrow cases.

**Comment by CyrusNajmabadi** [comment]
It's not clear to me that there's actually an issue.  The compiler tells me this:

![image](https://user-images.githubusercontent.com/4564579/77839622-44d06400-7133-11ea-9725-f2f1e8afb009.png)

Just by reading through the individual items i can see what the compiler is telling me.  And this item is particularly clear.  I'm using 'x' and it's hiding an outer variable.  I can fix that by no longer using 'x'.

**Comment by VBAndCs** [comment]
I tried this:
```
        Foo(Function() 3,
                Sub(x)
                    Dim y As Integer = New Exception()
                End Sub
        )
```

The error in the line:
`Dim y As Integer = New Exception()`
is not highlighted too, and the `Foo` is because the overload resolution .

I tried this in C#:
```
            Foo(() => 3,
                    (x) =>
                    {
                        int y = new Exception();                       
                    }
```

and the `new Exception();` is highlighted with the error:
>CS0029	Cannot implicitly convert type 'System.Exception' to 'int'	

You can check how C# does it.
